File: A08 Dynamic Programming/213. House Robber II/DP solution.py
###########
Time Complexity:
Space Complexity:


#############
class Solution:
    def rob(self, nums: List[int]) -> int:
        if (len(nums) == 0): 
            return 0
        if (len(nums) == 1):
            return nums[0]
        if (len(nums) == 2):
            return max(nums)
        
        dp1 = [0]*len(nums) 
        dp2 = [0]*len(nums)
        
        #start from 1st house 
        dp1[0], dp1[1] = nums[0], max(nums[0], nums[1])
        for i in range(2, len(nums) -1):
            dp1[i] = max(nums[i] + dp1[i-2] , dp1[i-1])
            print(dp1[i])
         
       #start from second house
        dp2[0], dp2[1] = 0, nums[1]
        for j in range(2, len(nums)):

            dp2[j] = max(nums[j] + dp2[j-2] , dp2[j-1])
        print(dp1[len(nums) -2])
        print(dp2[len(nums) -1])
        return max(dp1[len(nums) -2], dp2[len(nums) -1])
            
        

File: A08 Dynamic Programming/213. House Robber II/DP solution.java
/*
Time Complexity:
Space Complexity:

*/ 

int len = nums.length;
        if (len == 0)
            return 0;
        if (len < 2)
            return nums[0];

        int[] startFromFirstHouse = new int[len];
        int[] startFromSecondHouse = new int[len];

        startFromFirstHouse[0] = nums[0];
        startFromFirstHouse[1] = Math.max(nums[0], nums[1]);
        
        startFromSecondHouse[0] = 0;
        startFromSecondHouse[1] = nums[1];

        for (int i = 2; i < len - 1; i++) {
            startFromFirstHouse[i] = Math.max(startFromFirstHouse[i - 1], startFromFirstHouse[i - 2] + nums[i]);
        }
        for (int i = 2; i < len; i++) {
            startFromSecondHouse[i] = Math.max(startFromSecondHouse[i - 1], startFromSecondHouse[i - 2] + nums[i]);
        }
//        System.out.println(Arrays.toString(startFromFirstHouse));
//        System.out.println(Arrays.toString(startFromSecondHouse));

        return Math.max(startFromFirstHouse[len - 2], startFromSecondHouse[len - 1]);